File: IPaMusicPlayer/IPaMusicPlayer.m
//
//  IPaMusicPlayer.m
//  IPaMusicPlayer
//
//  Created by IPaPa on 13/9/7.
//  Copyright 2013年 IPaPa. All rights reserved.
//

#import "IPaMusicPlayer.h"



@interface IPaMusicPlayer() <AVAudioSessionDelegate,AVAudioPlayerDelegate>
@property (nonatomic,readonly) AVPlayer *musicPlayer;
@property (nonatomic,assign) BOOL shouldResume;
@property (nonatomic,assign) BOOL isPlay;
@end
@implementation IPaMusicPlayer
{
    
}
+(IPaMusicPlayer*)defaultPlayer
{
    static IPaMusicPlayer *defaultIPaMusicPlayer = nil;
    if (defaultIPaMusicPlayer == nil) {
        defaultIPaMusicPlayer = [[IPaMusicPlayer alloc] init];
    }
    return defaultIPaMusicPlayer;
}
-(id)init
{
    self = [super init];
    self.shouldResume = NO;
    //initial AudioSession
    // Registers this class as the delegate of the audio session to listen for audio interruptions
    [[AVAudioSession sharedInstance] setDelegate: self];
    //Set the audio category of this app to playback.
    NSError *setCategoryError = nil;
    [[AVAudioSession sharedInstance] setCategory: AVAudioSessionCategoryPlayback error: &setCategoryError];
    if (setCategoryError) {
        //RESPOND APPROPRIATELY
    }
    
    
    //Activate the audio session
    NSError *activationError = nil;
    [[AVAudioSession sharedInstance] setActive: YES error: &activationError];
    if (activationError) {
        //RESPOND APPROPRIATELY
    }
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(playerItemDidReachEnd:) name:AVPlayerItemDidPlayToEndTimeNotification object:nil];
    return self;
}
-(void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

/*-(void)setMusicPlayItems:(NSArray*)playerItems
 {
 [self.musicPlayer removeAllItems];
 self.musicPlayer = nil;
 self.musicPlayer = [AVQueuePlayer queuePlayerWithItems:playerItems];
 }*/
-(void)setMusicPlayItem:(AVPlayerItem*)playerItem
{
    if (_musicPlayer == nil)
    {
        _musicPlayer = [AVPlayer playerWithPlayerItem:playerItem];//[AVQueuePlayer playerWithPlayerItem:playerItem];
    }
    else {
        if (_musicPlayer.currentItem == playerItem) {
            [_musicPlayer seekToTime:CMTimeMake(0, 1)];
        }
        else {
            [_musicPlayer replaceCurrentItemWithPlayerItem:playerItem];
        }
        
        //        [self.musicPlayer insertItem:playerItem afterItem:nil];
        
    }
    
}
-(void)setMusicURL:(NSURL*)musicURL
{
    AVURLAsset *asset = [AVURLAsset assetWithURL:musicURL];
    AVPlayerItem *item = [AVPlayerItem playerItemWithAsset:asset];
    [self setMusicPlayItem:item];
    
}

-(void)pause
{
    self.isPlay = NO;
    [_musicPlayer pause];
}
-(void)play
{
    self.isPlay = YES;
    [_musicPlayer play];
}


- (void)playerItemDidReachEnd:(NSNotification *)notification {
    //allow for state updates, UI changes
    //    NSLog(@"song over!");
    self.isPlay = NO;
    [[NSNotificationCenter defaultCenter] postNotificationName:IPAMUSICPLAYER_ITEM_FINISHED object:self userInfo:nil];
}

#pragma mark - Global

+(void)setMusicURL:(NSURL*)musicURL
{
    [[IPaMusicPlayer defaultPlayer] setMusicURL:musicURL];
}
+(void)setMusicPlayItem:(AVPlayerItem*)playerItem
{
    [[IPaMusicPlayer defaultPlayer] setMusicPlayItem:playerItem];
}
/*+(void)setMusicPlayItems:(NSArray*)playerItems
 {
 [[IPaMusicPlayer defaultPlayer] setMusicPlayItems:playerItems];
 }*/
+(void)pause
{
    [[IPaMusicPlayer defaultPlayer] pause];
}
+(void)play
{
    [[IPaMusicPlayer defaultPlayer] play];
}
+(double)currentMusicDuration
{
    AVPlayerItem *currentItem = [IPaMusicPlayer currentItem];
    return (currentItem == nil)?0:CMTimeGetSeconds(currentItem.asset.duration);
    
}
+(double)currentMusicProgress
{
    AVPlayerItem *currentItem = [IPaMusicPlayer currentItem];
    return (currentItem == nil)?0:CMTimeGetSeconds(currentItem.currentTime);
}
+(AVPlayerItem*)currentItem
{
    return [IPaMusicPlayer defaultPlayer].musicPlayer.currentItem;
}

+(BOOL)isPlaying
{
    return [IPaMusicPlayer defaultPlayer].isPlay;
}
+(void)seekToTime:(NSUInteger)time
{
    AVPlayer *player = [IPaMusicPlayer defaultPlayer].musicPlayer;
    [player seekToTime:CMTimeMake(time, 1)];
}
#pragma mark - AVAudioSessionDelegate

/* the interruption is over */
- (void)endInterruptionWithFlags:(NSUInteger)flags
{
    if (self.shouldResume) {
        
        [self play];
        self.shouldResume = NO;
    }
    
}


- (void)beginInterruption /* something has caused your audio session to be interrupted */
{
    if(self.isPlay) {
        [self pause];
        self.shouldResume = YES;
        
    }
}

@end

File: IPaMusicPlayer/Classes/IPaMusicPlayer.swift
//
//  IPaMusicPlayer.swift
//  IPaMusicPlayer
//
//  Created by IPa Chen on 2019/1/3.
//

import UIKit
import AVFoundation

public class IPaMusicPlayer: NSObject {
    public static let shared = IPaMusicPlayer()
    var shouldResume:Bool = false
    var _isPlay:Bool = false
    public var timeControlStatus:AVPlayer.TimeControlStatus? {
        get {
            return self.musicPlayer?.timeControlStatus
        }
    }
    public static let IPaMusicPlayerItemFinished: NSNotification.Name = NSNotification.Name("IPaMusicPlayerItemFinished")
    public static let IPaMusicPlayerTimeStatusChanged:NSNotification.Name = NSNotification.Name("IPaMusicPlayerTimeStatusChanged")
    
    var currentItem:AVPlayerItem? {
        get {
            return self.musicPlayer?.currentItem
            
        }
    }
    public var playingUrl:URL? {
        get {
            return (self.currentItem?.asset as? AVURLAsset)?.url
        }
    }
    public var musicDuration:Double {
        get {
            if let currentItem = self.currentItem {
                return CMTimeGetSeconds(currentItem.currentTime())
            }
            return 0
        }
    }
    
    public var isPlay:Bool {
        get {
            return _isPlay
        }
    }
    fileprivate var _musicPlayer:AVPlayer?
    var musicPlayer:AVPlayer?
    {
        get {
            return _musicPlayer
        }
        set {
            if let _musicPlayer = _musicPlayer {
                _musicPlayer.removeObserver(self, forKeyPath: "timeControlStatu")
            }
            if let newValue = newValue {
                newValue.addObserver(self, forKeyPath: "timeControlStatus", options: [.old, .new], context: nil)
            }
            _musicPlayer = newValue
        }
    }
    public var currentAVMetadataItem:[AVMetadataItem] {
        get {
            return self.currentItem?.asset.commonMetadata ?? [AVMetadataItem]()
        }
    }
    public override init() {
        super.init()
        do {
            try AVAudioSession.sharedInstance().setCategory(.playback, mode: .default)
            try AVAudioSession.sharedInstance().setActive(true)
        } catch let error {
            print(error)
        }
        NotificationCenter.default.addObserver(self, selector: #selector(onPlayerItemDidReachEnd(_:)), name: .AVPlayerItemDidPlayToEndTime, object: nil)
        
        NotificationCenter.default.addObserver(self, selector: #selector(onAudioSessionInterruption(_:)) ,name: AVAudioSession.interruptionNotification, object: nil)
        
    }
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    func setMusicPlayItem(_ playerItem:AVPlayerItem) {
        if let musicPlayer = musicPlayer {
            if musicPlayer.currentItem == playerItem {
                musicPlayer.seek(to: CMTime(value: 0, timescale: 1))
            }
            else {
                musicPlayer.replaceCurrentItem(with: playerItem)
            }
        }
        else {
            musicPlayer = AVPlayer(playerItem: playerItem)
        }
        
    }
    public func setMusicURL(_ musicURL:URL) {
        let asset = AVURLAsset(url: musicURL)
        let item = AVPlayerItem(asset: asset)
        self.setMusicPlayItem(item)
    }
    
    public func pause() {
        self._isPlay = false
        self.musicPlayer?.pause()
    }
    public func play() {
        self._isPlay = true
        self.musicPlayer?.play()
    }
    public func seekToTime(_ time:Int) {
        guard let musicPlayer = musicPlayer else {
            return
        }
        musicPlayer.seek(to: CMTime(value: CMTimeValue(time), timescale: 1))
    }
    @objc func onPlayerItemDidReachEnd(_ notification:Notification) {
        self._isPlay = false
        NotificationCenter.default.post(name: IPaMusicPlayer.IPaMusicPlayerItemFinished, object: self)
    }
    @objc func onAudioSessionInterruption(_ notification:Notification) {
        guard let interruptionType = notification.userInfo?[AVAudioSessionInterruptionTypeKey] as? AVAudioSession.InterruptionType else {
            return
            
        }
        
        switch interruptionType {
            
        case .began:
            if self.isPlay {
                self.pause()
                self.shouldResume = true
            }
        case .ended:
            if self.shouldResume {
                self.play()
                
                self.shouldResume = false
            }
        @unknown default:
            break
        }
        
    }
    override public func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if keyPath == "timeControlStatus", let change = change, let newValue = change[NSKeyValueChangeKey.newKey] as? Int, let oldValue = change[NSKeyValueChangeKey.oldKey] as? Int {
            let oldStatus = AVPlayer.TimeControlStatus(rawValue: oldValue)
            let newStatus = AVPlayer.TimeControlStatus(rawValue: newValue)
            if newStatus != oldStatus {
                NotificationCenter.default.post(name: IPaMusicPlayer.IPaMusicPlayerTimeStatusChanged, object: nil)
            }
        }
    }
}
